### PR TITLE
feat(client): pass metrics context metadata to the back end

### LINKED ISSUE #3551: Propagate event metadata to the auth server
To facilitate https://github.com/mozilla/fxa-auth-server/issues/1200 and [FxA-70](https://github.com/mozilla/fxa/blob/master/features/FxA-70-kpi-dashboards/README.md#data-capture-and-storage), the front-end must propagate event metadata to the auth server.

Specifically, the existing fields `service`, `context`, `entrypoint`, `campaign` and `migration` are required, as is the new `flowId` that will be created as part of #3550.

Agreement is obviously needed with the auth server as to whether they should be sent in the query string or in the request body.

### PR DESCRIPTION
**Pull request by philbooth**
Fixes #3551.

Adds metrics context metadata to the auth server `/signup` and `/login` requests. `flowId` and `flowBeginTime` are always set, other fields are set according to their presence on the relier object.

@vladikoff r?

### CODE REVIEW

**Comment by philbooth** [comment]
Example request data, in case the description was unclear:

screen shot 2016-05-03 at 10 35 52

screen shot 2016-05-03 at 11 51 09

File: app/tests/spec/lib/fxa-client.js [hunk 1]
```diff
@@ -458,6 +480,27 @@ define(function (require, exports, module) {
...
+          return p({});
+        });
+
+        relier.set('service', 'chronicle');
+
+        return client.signIn(email, password, relier, {
+          metricsContext: { foo: 'bar' }
+        })
+        .then(function () {
+          console.dir(realClient.signIn.args[0]);
```
Comment (vladikoff) [inline app/tests/spec/lib/fxa-client.js]: 👀

**Comment by vladikoff** [comment]
@philbooth hm could you please try to rebase / repush this PR? For some reason circle ci does not like it... I pushed my own PR and it is running okay here: https://circleci.com/gh/mozilla/fxa-content-server/2926

**Comment by vladikoff** [comment]
> @philbooth hm could you please try to rebase / repush this PR? For some reason circle ci does not like it... I pushed my own PR and it is running okay here: https://circleci.com/gh/mozilla/fxa-content-server/2926

Hm it could be because the remote circle ci server is outdated and does not support flow id?

**Comment by philbooth** [comment]
> Hm it could be because the remote circle ci server is outdated and does not support flow id?

Oh, possibly. It was part of auth server tag `v1.59.0`, what is circle running?

**Comment by vladikoff** [comment]
@philbooth running 

> node_modules/.bin/intern-runner config=tests/intern_functional_circle fxaAuthRoot=http://fxadeploybox.dev.lcip.org:9000 fxaEmailRoot=http://fxadeploybox.dev.lcip.org:9001 fxaOauthApp=http://fxadeploybox.dev.lcip.org:8080 fxaUntrustedOauthApp=http://fxadeploybox.dev.lcip.org:10139 fxaProduction=true


locally shows that is is failing to create an account it seems. Thoughts?

**Comment by vladikoff** [comment]
I updated the fxa-auth-server on the deploy box to latest

```
commit f43607189407f3453456768ead8e7668825ac730
Author: Ubuntu <>
Date:   Tue Jan 26 16:11:10 2016 +0000

    remote configuration

commit 2869dc129f76b6abbf0cf4a4aca9c26340d7ca9d
Merge: e5377a3 af748be
Author: Ryan Kelly <ryan@rfk.id.au>
Date:   Tue May 3 07:07:53 2016 +1000

    Merge pull request #1245 from mozilla/device-update-safety-switch
    
    feat(devices): Add metrics on device updates, and a flag to disable them
```

**Comment by philbooth** [comment]
Closing this, there are a couple of things wrong with it.

**Comment by philbooth** [comment]
@vladikoff, I've pushed b237ee7 to fix yesterday's functional test failures, apologies for those again.

Can I get an r?

**Comment by philbooth** [comment]
Hang on, still getting some functional test failures locally, looking into them now.

**Comment by shane-tomlinson** [comment]
> Hang on, still getting some functional test failures locally, lookin

I'm on the case - https://github.com/mozilla/fxa-content-server/issues/3704 and https://github.com/mozilla/fxa-content-server/pull/3706

**Comment by philbooth** [comment]
Man, I am such a fool sometimes. This functional tests are fine, the code is fine, this PR is fine. Some local auth server changes, which I'd completely forgotten about, were messing with my shizzle.

@vladikoff, once again, r?

**Comment by vladikoff** [comment]
From IRC discussion: let's try to see if making the new `metrics-context` be a model with the `lib/metrics`. 
That might help simplify the `app-start` functionality and keep the metrics stuff more self-contained. 
some notes:
* there is also a 'relier' `_context` in metrics, so we might need to do something to avoid confusion
* some relier fields and methods can be reused in both modules.

**Comment by philbooth** [comment]
@vladikoff, not sure whether it's what you had in mind but I've pushed 4cec4e4 to move the `metricsContext` model into the `metrics` object.

It is renamed to `ActivityEventMetadata` and instantiated as `this._activityEventMetadata` to avoid confusion with the existing `this._context` property. I've also exposed two new methods on the `metrics` object, `getActivityEventMetadata` and `setActivityEventMetadata`, for interacting with it.

Otherwise, it should be pretty much the same as what I had before. r?

**Comment by vladikoff** [comment]
This is great. Awesome work!

**Pull request merged**